r/src/expression.cpp

```diff
@@ -21,99 +21,97 @@
 
 // [[arrow::export]]
 std::shared_ptr<ds::Expression> dataset___expr__field_ref(std::string name) {
-  return std::make_shared<ds::FieldExpression>(std::move(name));
+  return ds::field_ref(std::move(name));
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::ComparisonExpression> dataset___expr__equal(
+std::shared_ptr<ds::Expression> dataset___expr__equal(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<ds::Expression>& rhs) {
   return ds::equal(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::ComparisonExpression> dataset___expr__not_equal(
+std::shared_ptr<ds::Expression> dataset___expr__not_equal(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<ds::Expression>& rhs) {
   return ds::not_equal(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::ComparisonExpression> dataset___expr__greater(
+std::shared_ptr<ds::Expression> dataset___expr__greater(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<ds::Expression>& rhs) {
   return ds::greater(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::ComparisonExpression> dataset___expr__greater_equal(
+std::shared_ptr<ds::Expression> dataset___expr__greater_equal(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<ds::Expression>& rhs) {
   return ds::greater_equal(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::ComparisonExpression> dataset___expr__less(
+std::shared_ptr<ds::Expression> dataset___expr__less(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<ds::Expression>& rhs) {
   return ds::less(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::ComparisonExpression> dataset___expr__less_equal(
+std::shared_ptr<ds::Expression> dataset___expr__less_equal(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<ds::Expression>& rhs) {
   return ds::less_equal(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<ds::InExpression> dataset___expr__in(
+std::shared_ptr<ds::Expression> dataset___expr__in(
     const std::shared_ptr<ds::Expression>& lhs,
     const std::shared_ptr<arrow::Array>& rhs) {
-  return std::make_shared<ds::InExpression>(lhs->In(rhs));
+  return lhs->In(rhs).Copy();
```

Comment:
Why does this require `.Copy()`? Would it be clearer if there were `ds::in_(lhs, rhs)` like the other operators?

Reply:
`Copy()` moves the expression from the stack into a shared_ptr, replacing make_shared.

I can add them if desired.

Reply:
🤷 not a big deal to me